onnxruntime/core/providers/cpu/object_detection/non_max_suppression.cc

```diff
@@ -24,192 +25,171 @@ ONNX_OPERATOR_KERNEL_EX(
+      std::priority_queue<ScoreIndexPair, std::deque<ScoreIndexPair>> sorted_scores_with_index;
+      const auto* class_scores = scores_data + box_score_offset;
+      if (pc.score_threshold_ != nullptr) {
+        for (int64_t box_index = 0; box_index < pc.num_boxes_; ++box_index, ++class_scores) {
+          if (*class_scores > score_threshold) {
+            sorted_scores_with_index.emplace(*class_scores, box_index);
+          }
+        }
+      } else {
+        for (int64_t box_index = 0; box_index < pc.num_boxes_; ++box_index, ++class_scores) {
+          sorted_scores_with_index.emplace(*class_scores, box_index);
         }
       }
 
       ScoreIndexPair next_top_score;
       std::vector<int64_t> selected_indicies_inside_class;
-      // Get the next box with top score, filter by iou_threshold_
+      // Get the next box with top score, filter by iou_threshold
       while (!sorted_scores_with_index.empty()) {
         next_top_score = sorted_scores_with_index.top();
         sorted_scores_with_index.pop();
 
         bool selected = true;
         // Check with existing selected boxes for this class, suppress if exceed the IOU (Intersection Over Union) threshold
-        for (int64_t selected_indicies_inside_clas : selected_indicies_inside_class) {
-          if (SuppressByIOU(boxes_data + box_offset, selected_indicies_inside_clas, next_top_score.index,
-                            iou_threshold)) {
+        for (int64_t selected_index : selected_indicies_inside_class) {
+          if (SuppressByIOU(boxes_data + box_offset, selected_index, next_top_score.index_,
+                            center_point_box, iou_threshold)) {
             selected = false;
             break;
           }
         }
 
         if (selected) {
-          if (max_output_boxes_per_class > 0 && static_cast<int64_t>(selected_indicies_inside_class.size()) >= max_output_boxes_per_class) {
+          if (max_output_boxes_per_class > 0 &&
+              static_cast<int64_t>(selected_indicies_inside_class.size()) >= max_output_boxes_per_class) {
             break;
           }
-          selected_indicies_inside_class.push_back(next_top_score.index);
-          tmp_selected_indices.emplace_back(batch_index, class_index, next_top_score.index);
+          selected_indicies_inside_class.push_back(next_top_score.index_);
+          selected_indices.emplace_back(batch_index, class_index, next_top_score.index_);
         }
       }  //while
     }    //for class_index
   }      //for batch_index
 
-  auto num_selected = static_cast<int32_t>(tmp_selected_indices.size());
-  Tensor* selected_indices = ctx->Output(0, {num_selected, 3});
-  ORT_ENFORCE(selected_indices);
-  memcpy(selected_indices->MutableData<int64_t>(), tmp_selected_indices.data(), num_selected * sizeof(selected_index));
+  const auto num_selected = selected_indices.size();
+  Tensor* output = ctx->Output(0, {static_cast<int64_t>(num_selected), 3});
+  ORT_ENFORCE(output != nullptr);
```

Comment:
Either use gsl::span (output->MutableDataAsSpan) with gsl::copy instead of memcpy, or validate the '3' against sizeof(SelectedIndex) so that the code wouldn't silently break if someone added something to SelectedIndex.

Reply:
This is a very good point.

---
In reply to: [298065552](https://github.com/microsoft/onnxruntime/pull/1303#discussion_r298065552) [](ancestors = 298065552)